src/db_impl.cc

```diff
@@ -175,114 +176,140 @@ Status TitanDBImpl::ValidateOptions(
-                            MutableTitanCFOptions(descs[i].options),
-                            base_table_factory, titan_table_factory}));
-      base_descs[i].options.table_factory = titan_table_factory;
-      // Add TableProperties for collecting statistics GC
-      base_descs[i].options.table_properties_collector_factories.emplace_back(
-          std::make_shared<BlobFileSizeCollectorFactory>());
-    }
-    handles->clear();
-    s = db_->Close();
-    delete db_;
-    db_ = nullptr;
-  }
-  if (!s.ok()) return s;
-
-  if (stats_.get()) {
-    stats_->Initialize(column_families);
-  }
-
-  s = blob_file_set_->Open(column_families);
-  if (!s.ok()) return s;
-
   // Initialize GC thread pool.
   if (!db_options_.disable_background_gc && db_options_.max_background_gc > 0) {
     env_->IncBackgroundThreadsIfNeeded(db_options_.max_background_gc,
                                        Env::Priority::BOTTOM);
   }
-
+  // Open base DB.
   s = DB::Open(db_options_, dbname_, base_descs, handles, &db_);
-  if (s.ok()) {
-    db_impl_ = reinterpret_cast<DBImpl*>(db_->GetRootDB());
-    ROCKS_LOG_INFO(db_options_.info_log, "Titan DB open.");
-    ROCKS_LOG_HEADER(db_options_.info_log, "Titan git sha: %s",
-                     titan_build_git_sha);
-    db_options_.Dump(db_options_.info_log.get());
-    for (auto& desc : descs) {
-      ROCKS_LOG_HEADER(db_options_.info_log,
-                       "Column family [%s], options:", desc.name.c_str());
-      desc.options.Dump(db_options_.info_log.get());
+  if (!s.ok()) {
+    db_ = nullptr;
+    handles->clear();
+    return s;
+  }
+  db_impl_ = reinterpret_cast<DBImpl*>(db_->GetRootDB());
+  assert(db_ != nullptr);
+  assert(handles->size() == descs.size());
+  std::map<uint32_t, TitanCFOptions> column_families;
+  std::vector<ColumnFamilyHandle*> cf_with_compaction;
+  for (size_t i = 0; i < descs.size(); i++) {
+    cf_info_.emplace((*handles)[i]->GetID(),
+                     TitanColumnFamilyInfo(
+                         {(*handles)[i]->GetName(),
+                          ImmutableTitanCFOptions(descs[i].options),
+                          MutableTitanCFOptions(descs[i].options),
+                          descs[i].options.table_factory /*base_table_factory*/,
+                          titan_table_factories[i]}));
+    column_families[(*handles)[i]->GetID()] = descs[i].options;
+    if (!descs[i].options.disable_auto_compactions) {
+      cf_with_compaction.push_back((*handles)[i]);
     }
-  } else {
-    ROCKS_LOG_ERROR(db_options_.info_log, "Titan DB open failed: %s",
-                    s.ToString().c_str());
+  }
+  // Initialize Titan internals.
+  if (stats_ != nullptr) {
+    stats_->Initialize(column_families);
+  }
+  s = blob_file_set_->Open(column_families);
+  if (!s.ok()) {
+    return s;
+  }
+  TEST_SYNC_POINT_CALLBACK("TitanDBImpl::OpenImpl:BeforeInitialized", db_);
+  // Initialization done.
+  initialized_ = true;
```

Comment:
seems `initialized_` should be set in background later?

Reply:
No. After `initialized_=true` is set, compaction and flush can generate blob files. After the GC stats patch (which I'm working on), GC stats will be initialize after this point, in background, and after that GC job will start.